contracts/test/signatureBased.test.ts

```diff
@@ -0,0 +1,289 @@
+import { expect } from "chai";
+import { Wallet, Provider, Contract, utils } from "zksync-web3";
+import hardhatConfig from "../hardhat.config";
+import { Deployer } from "@matterlabs/hardhat-zksync-deploy";
+import * as ethers from "ethers";
+
+import { deployContract, fundAccount, setupDeployer } from "./utils";
+
+import dotenv from "dotenv";
+import { _TypedDataEncoder } from "ethers/lib/utils";
+dotenv.config();
+
+const PRIVATE_KEY =
+  process.env.WALLET_PRIVATE_KEY ||
+  "0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110";
+const abiCoder = new ethers.utils.AbiCoder();
+
+describe("SignatureBasedPaymaster", function () {
+  let provider: Provider;
+  let wallet: Wallet;
+  let deployer: Deployer;
+  let userWallet: Wallet;
+  let signerWallet: Wallet;
+  let paymaster: Contract;
+  let greeter: Contract;
+
+  before(async function () {
+    const deployUrl = hardhatConfig.networks.zkSyncInMemory.url;
+    [provider, wallet, deployer] = setupDeployer(deployUrl, PRIVATE_KEY);
+    const emptyWallet = Wallet.createRandom();
+    console.log(`User wallet's address: ${emptyWallet.address}`);
+    userWallet = new Wallet(emptyWallet.privateKey, provider);
+    signerWallet = Wallet.createRandom();
+    console.log(`Signer wallet's address: ${signerWallet.address}`);
+    signerWallet = new Wallet(signerWallet.privateKey, provider);
+    paymaster = await deployContract(deployer, "SignatureBasedPaymaster", [
+      signerWallet.address,
+    ]);
+    greeter = await deployContract(deployer, "Greeter", ["Hi"]);
+    await fundAccount(wallet, paymaster.address, "3");
+    console.log(`Paymaster current signer: ${signerWallet.address}`);
+  });
+
+  async function createSignatureData(
+    signer: Wallet,
+    user: Wallet,
+    delay: number,
```

Comment:
Rather then using `delay`, and specifically for testing expired signatures consider making use of: 

`evm_setNextBlockTimestamp` and `evm_mine` instead. You can review the time based paymaster tests for an example. I think this would make the tests more straightforward.

Reply:
To be clear, these methods are exclusive to testing with `era-test-node`

Reply:
I am using those hardhat commands for signature expiry tests. 
https://github.com/matter-labs/paymaster-examples/blob/eca49e35e24c3989be1e528aa9ec600ff7614b30/contracts/test/signatureBased.test.ts#L173-L174

Perhaps, the variable name `delay` is causing confusion. This variable basically is the number of seconds after the current block.timestamp post which the signature expired. I will rename those variable to "validSeconds" or something similar. 

I have noted rest of the changes and will push them soon.